Rest: Handle IPv6 link local addresses with zone identifier in Authenticate

rapour opened the pull request:
This PR suggests a fix for #268 

Closes https://github.com/canonical/microcluster/issues/268

roosterfish commented on internal/db/dqlite.go:
```diff
@@ -397,7 +398,7 @@ func dqliteNetworkDial(ctx context.Context, addr string, db *DqliteDB) (net.Conn
 	}
 
-	path := fmt.Sprintf("https://%s/%s/%s", addr, internalTypes.InternalEndpoint, "database")
+	path := fmt.Sprintf("https://%s/%s/%s", strings.Replace(addr, "%", "%25", 1), internalTypes.InternalEndpoint, "database")
```
The `url.Parse` (and the internal [`parse`](https://github.com/golang/go/blob/master/src/net/url/url.go)) do have some logic checks for this but as you have mentioned they expect the zone identifier to be already encoded (check for `%25`).
So I wonder if we can circumvent this by crafting the `request.URL` in a different way. Meaning that we provide the host and URL separately. Maybe the `url` package already has some tooling that could be useful for this?

rapour replied:
I couldn't find anything useful from the `url` package as it probably doesn't recognize the textual representation of ipv6 in RFC4007:
https://github.com/golang/go/blob/b07b20fbb591ac77158e1089299ce5acad71ffde/src/net/url/url_test.go#L704-L709
If you had any other directions in mind I would be happy to investigate. However, we can think of an alternative too. My impression is that `addr` string here is of type `AddrPort`, since `NewCluster` parses the input address into `AddrPort`. If it is indeed the case, we could first  parse the `addr` string into `AddrPort` and then handle these subtleties there in a canonical manner.

roosterfish replied:
Is it your idea to modify the `AddrPort` type's `String()` to include the encoding? That sounds reasonable and also makes it usable for future use cases.

rapour replied:
I'm afraid changing the `String()` encoding may break other parts of the code that expect non-encoded `host%zone`. How do you feel about something like this: 
```
func dqliteNetworkDial(ctx context.Context, addr string, db *DqliteDB) (net.Conn, error) {
	addrPort, err := types.ParseAddrPort(addr)
	if err != nil {
		return nil, fmt.Errorf("Failed to parse the address: %w", err)
	}
	....
	request.URL = &url.URL{
		Scheme: "https",
		Host:   addrPort.String(),
		Path:   fmt.Sprintf("/%s/%s", internalTypes.InternalEndpoint, "database"),
	}
	...
}
```
This way we can actually circumvent the `url.Parse` and still keep the address validation.

roosterfish replied:
That's exactly what I meant in my original comment:

> So I wonder if we can circumvent this by crafting the request.URL in a different way. Meaning that we provide the host and URL separately.

So if I understood your comment right, we don't have to apply any special encoding when doing `Host: addrPort.String()`? That seems much preferable.

rapour replied:
Yes Exactly

roosterfish commented on rest/types/addrport.go:
```diff
@@ -94,6 +94,19 @@ func (a AddrPort) String() string {
 
+// WithoutZone returns the AddrPort with the zone identifier removed from the address. It will return the same
+// AddrPort as the receiver if the address is ipv4 or ipv6 without zone identifier.
+func (a AddrPort) WithoutZone() AddrPort {
```
It bugs me a little bit that there is nothing in the standard library for such checks so we have to remove the zone ID suffix ourselves for this check.

I think it's ok that we store the IPv6 with zone ID internally as this is what was ultimately configured so it's getting passed down in `hostAddress`. 
Have you checked if the `x/net/ipv6` package has something useful we can use to deal with the zone ID in the address?

rapour replied:
The approach here is to strip the zone ID so it can be compared to the host and it uses the following feature from the standard library: 

https://github.com/golang/go/blob/b07b20fbb591ac77158e1089299ce5acad71ffde/src/net/netip/netip.go#L701-L704

Although the check is manual, the way to strip the zone ID is something done via the standard library. Interestingly, I found another way to strip the zone ID by using `WithZone("")`:
https://github.com/golang/go/blob/b07b20fbb591ac77158e1089299ce5acad71ffde/src/net/netip/netip.go#L489-L492
We can move to this one but at the end, both do the same and strip the zone ID so it becomes comparable to `Host`. The `netip` package has some comparison methods but I suspect any other solution would be as mediocre as this one since http request's `r.Host` is a string and will need parsing before any other checks. The `RemoteAddr` of the http request also has the full `host%zone` info which is comparable to the original `hostAddress` but there are no guarantees about the format of it and I'm not sure whether it's a good idea to replace it with the `r.Host`. 

As a side note, please also [see this](https://github.com/golang/go/issues/71362).

roosterfish replied:
I like the idea of using the standard libraries `WithZone("")` using an empty string. Let's do this.

rapour replied:
Done, now we have the same interface for `types.AddrPort` as there is for `netip.Addr` when it comes to zone ID modifications.

roosterfish replied:
Can we use the `WithZone("")` from the std lib? I thought we use this one instead of having our own `WithoutZone()`.

roosterfish replied:
Saw the latest commits too late. Fixed.

roosterfish reviewed (commented):
Hi, thanks for your investigation on this. Please see my comments below.

roosterfish commented:
Can you please rebase from `v3`. We have added some more linter rules.

roosterfish commented on rest/types/addrport_test.go:
```diff
@@ -8,7 +8,6 @@ import (
 )
 
 func TestAddrPortWithoutZone(t *testing.T) {
-
```
We don't squash the commits when merging. Can you please squash this commit with the one where you have introduced the test?
This way you don't have a commit in the PR which adds lines which are being removed again by one of the follow up commits.

roosterfish replied:
This also applies to the other latest changes regarding `WithZone("")`.

rapour replied:
Done

roosterfish commented:
Ah sorry. I haven't seen that you have already all the changes in the last commits.
Please see my other comment regarding squashing.

roosterfish reviewed (dismissed):
Thanks for this!

roosterfish commented:
@rapour I cannot merge before your commits aren't signed as this is a rule of the repo.

rapour committed:
internal/db: fix invalid argument error when parsing ipv6 with zone identifier

net.url package conforms to rfc6874\nwe need to first parse the addr to an AddrPort struct and then craft the request URL directly

Signed-off-by: Reza Abbasalipour <reza.abbasalipour@canonical.com>

rapour committed:
rest/types: add a method to modify zone identifier of AddrPort

netip provides a WithZone utility for netip.Addr to modify the zone identifier\nthe WithZone receiver method defined here provides the same utility for the address part of the AddrPort

Signed-off-by: Reza Abbasalipour <reza.abbasalipour@canonical.com>

rapour committed:
rest/access: exclude zone identifier from address in authentication

the authentication function compares the http request host with address host to make some decisions which\nis undermined when zone identifiers are present in ipv6 addresses\nexcluding the zone identifier will solve the problem \nthe result will not change in case of ipv4 or ipv6 without zone identifier

Signed-off-by: Reza Abbasalipour <reza.abbasalipour@canonical.com>

rapour committed:
internal/rest/resources: fix linting error

Signed-off-by: Reza Abbasalipour <reza.abbasalipour@canonical.com>

roosterfish reviewed (approved):
LGTM!
